src/devicecontroller/DefaultDeviceController.ts

```diff
@@ -555,6 +555,10 @@ export default class DefaultDeviceController implements DeviceControllerBasedMed
           return (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters)
             .exact as string;
         }
+      } else {
+        if (deviceMediaTrackConstraints.deviceId as string) {
```

Comment:
I understand you followed the pattern in this function. If I'm not missing TypeScript features, using `as` in `if` doesn't have any effect. @RidipDe 

[Click here to try in the TS playgrond](https://www.typescriptlang.org/play/?target=5&ssl=1&ssc=1&pln=7&pc=2#code/JYMwBAFA5FYIYGcwIC4CdgDsDmBKMA3gFBilgDGA9pgpQDYCmAdHZdtAHKUphoNzkAFgwAmUXAG4iAXyJFQkFAE8ADg0rgYYALy6wUVBhzjCJMlRr1mrdlAAqg4EifwKlNH3I8A7nCXiZIA):
```
if ('' as string) {
    console.log('Not reached');
}

if (typeof '' === 'string') {
    console.log('This is a correct way')
}
```

Reply:
I read about union type in typescript here: https://www.typescriptlang.org/docs/handbook/advanced-types.html
and implemented accordingly. Its not just checking the type in the above implementation. Union types are useful for modeling situations when values can overlap in the types they can take on.

deviceMediaTrackConstraints.deviceId can be either
``` string | string[] | ConstrainDOMStringParameters```

The scenario already handled is if deviceMediaTrackConstraints.deviceId  is referred as ConstrainDOMStringParameters then get the string `exact` attribute. 

We have just handled this condition as of now because in our SDK, we are currently assigning the device id as a `string` as an `exact` property in this line https://github.com/aws/amazon-chime-sdk-js/blob/master/src/devicecontroller/DefaultDeviceController.ts#L731

We cannot do 
```
if (typeof deviceMediaTrackConstraints.deviceId === ConstrainDOMStringParameters)

as 'ConstrainDOMStringParameters' only refers to a type and not value
```

As per need we need to keep adding the checks
example against a string and string[] and ideal
For checking if its type of 'string' or not we need to use this
```
if (typeof ' ' === 'string') {
    console.log('This is a correct way')
}
```


If we want the `ideal` string property of the device, we need to add
```
      /* istanbul ignore else */
      if ((deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal) {
        /* istanbul ignore else */
        if (
          (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal as string
        ) {
          return (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters)
            .ideal as string;
        }
      }
```

Similarly, as per the need, we need to handle string[] for properties `exact` and `ideal`

Reply:
> If we want the ideal string property of the device, we need to add

You might be confused of the type conversion (`as string`) vs. the type check. Assume that `deviceId` is a string array. The following code will still return a string array.

```
/* istanbul ignore else */
if ((deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal) {
  /* istanbul ignore else */
  if (
    (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal as string
  ) {
    return (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters)
      .ideal as string;
  }
}
```

The 2nd `if` statement doesn't check whether the condition is a string or not, but it will be `true` if the condition is a non-falsy item. i.e. It's no different from having only one `if` statement.

```
if ((deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal) {
  return (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters)
    .ideal as string;
}
```

To check if the `ideal` type is a string, you can use `typeof`.
```
if (typeof (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal === 'string') {
  return (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters)
    .ideal as string;
}
```